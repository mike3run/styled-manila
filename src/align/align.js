import { css } from 'styled-components'

const horizontalAlignment = css`
  left: 50%;
  transform: translateX(-50%);
`

const verticalAlignment = css`
  top: 50%;
  transform: translateY(-50%);
`

const bothAlignments = css`
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`

/**
 * Center an element on either or both axis
 * @param {string} direction
 * @returns {string}
 */
const align = direction => {
  const horizontal = direction === 'horizontal'
  const vertical = direction === 'vertical'
  const both = direction === 'both'

  let alignmentNeeded = bothAlignments

  if (both) alignmentNeeded = bothAlignments
  if (horizontal) alignmentNeeded = horizontalAlignment
  if (vertical) alignmentNeeded = verticalAlignment

  return css`
    position: absolute;
    transform-style: preserve-3d;
    ${alignmentNeeded}
  `
}

export default align
