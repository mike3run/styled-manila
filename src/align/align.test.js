import React from 'react'
import styled from 'styled-components'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import align from './align'

const ExpectedDefaults = tree => {
  expect(tree).toHaveStyleRule('position', 'absolute')
  expect(tree).toHaveStyleRule('transform-style', 'preserve-3d')
}

test('Explicit Both Aligns Work!', () => {
  const Test = styled.div`
    ${align('both')}
  `
  const tree = renderer.create(<Test />).toJSON()
  expect(tree).toMatchSnapshot()
  ExpectedDefaults(tree)
  expect(tree).toHaveStyleRule('top', '50%')
  expect(tree).toHaveStyleRule('left', '50%')
  expect(tree).toHaveStyleRule('transform', 'translate(-50%,-50%)')
})

test('Implicit Both Aligns Work!', () => {
  const Test = styled.div`
    ${align()}
  `
  const tree = renderer.create(<Test />).toJSON()
  expect(tree).toMatchSnapshot()
  ExpectedDefaults(tree)
  expect(tree).toHaveStyleRule('top', '50%')
  expect(tree).toHaveStyleRule('left', '50%')
  expect(tree).toHaveStyleRule('transform', 'translate(-50%,-50%)')
})

test('Horizontal Aligns Work!', () => {
  const Test = styled.div`
    ${align('horizontal')}
  `
  const tree = renderer.create(<Test />).toJSON()
  expect(tree).toMatchSnapshot()
  ExpectedDefaults(tree)
  expect(tree).toHaveStyleRule('left', '50%')
  expect(tree).toHaveStyleRule('transform', 'translateX(-50%)')
})

test('Vertical Aligns Work!', () => {
  const Test = styled.div`
    ${align('vertical')}
  `
  const tree = renderer.create(<Test />).toJSON()
  expect(tree).toMatchSnapshot()
  ExpectedDefaults(tree)
  expect(tree).toHaveStyleRule('top', '50%')
  expect(tree).toHaveStyleRule('transform', 'translateY(-50%)')
})
