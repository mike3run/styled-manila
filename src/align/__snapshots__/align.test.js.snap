// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explicit Both Aligns Work! 1`] = `
.c0 {
  position: absolute;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

<div
  className="c0"
/>
`;

exports[`Horizontal Aligns Work! 1`] = `
.c0 {
  position: absolute;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  left: 50%;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

<div
  className="c0"
/>
`;

exports[`Implicit Both Aligns Work! 1`] = `
.c0 {
  position: absolute;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

<div
  className="c0"
/>
`;

exports[`Vertical Aligns Work! 1`] = `
.c0 {
  position: absolute;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

<div
  className="c0"
/>
`;
